bin/genesis

```diff
@@ -26,6 +26,192 @@ if ($ENV{BOSH_ALL_PROXY}) {
+		$prog .= ' "$@"'; # old style of passing in args as array, need to wrap for shell call
+	}
+
+	local %ENV = %ENV; # To get local scope for duration of this call
+	for (keys %{$opts{env} || {}}) {
+		$ENV{$_} = $opts{env}{$_};
+		main::debug("#M{Setting: }#B{$_}='#C{$ENV{$_}}'");
+	}
+	my $shell = $opts{shell} || '/bin/bash';
+	$prog .= ($opts{stderr} ? " 2>$opts{stderr}" : ' 2>&1') unless ($opts{interactive});
+	main::debug("#M{Executing:} `#C{$prog}`%s", ($opts{interactive} ? " #Y{(interacively)}" : ''));
+	if (@args) {
+		unshift @args, $shell;
+		main::debug("#m{ - with arguments:}");
+		main::debug("#m{%4s:} '#c{%s}'", $_, $args[$_]) for (1..$#args);
+	}
+
+	my @cmd = ($shell, "-c", $prog, @args);
+	my $out;
+	if ($opts{interactive}) {
+		system @cmd;
+	} else {
+		open my $pipe, "-|", @cmd;
+		$out = do { local $/; <$pipe> };
+		close $pipe;
+	}
+	my $rc = $? >>8;
+	if ($rc) {
+		die("#R{$opts{onfailure}}%s\n", defined($out) ? ":\n$out" :'') if $opts{onfailure};
+		main::debug("#R{==== ERROR: $rc}");
+		main::debug("$out\n#R{==== END}") if defined($out);
+	} else {
+		main::debug("#g{Command executed successfully.}");
+	}
+	return
+		$opts{passfail}    ? $rc == 0 :
+		$opts{interactive} ? $rc :
+		wantarray          ? ($out,$rc) :
+		$out;
+}
+
+sub interact {
+	my @args = @_;
+	my $opts = (ref($args[0]) eq 'HASH') ? shift @args : {};
+	$opts->{interactive} = 1;
+	$opts->{passfail} = 1;
+	return execute($opts,@args);
+}
+
+sub check {
+	my @args = @_;
+	my $opts = (ref($args[0]) eq 'HASH') ? shift @args : {};
+	$opts->{interactive} = 0;
+	$opts->{passfail} = 1;
+	return execute($opts,@args);
+}
+
+sub bosh {
+	my @args = @_;
+	my $opts = (ref($args[0]) eq 'HASH') ? shift @args : {};
+	$opts->{env} ||= {};
+	# Clear out the BOSH env vars unless we're under Concourse
+	unless ($ENV{BUILD_PIPELINE_NAME}) {
+		$opts->{env}{HTTPS_PROXY} = ''; # bosh dislikes this env var
+		$opts->{env}{https_proxy} = ''; # bosh dislikes this env var
+		$opts->{env}{BOSH_ENVIRONMENT}   ||= ''; # ensure using ~/.bosh/config via alias
+		$opts->{env}{BOSH_CA_CERT}       ||= '';
+		$opts->{env}{BOSH_CLIENT}        ||= '';
+		$opts->{env}{BOSH_CLIENT_SECRET} ||= '';
+	}
+	if ($args[0] =~ /\$\{?[\@0-9]/ && scalar(@args) > 1) {
+		$args[0] = main::bosh_cmd() ." $args[0]";
+	} else {
+		unshift @args, main::bosh_cmd()
+	}
+	return wantarray ? (execute($opts,@args)) : execute($opts,@args);
+}
+
+sub bosh_interact {
+	my @args = @_;
+	my $opts = (ref($args[0]) eq 'HASH') ? shift @args : {};
+	$opts->{interactive} = 1;
+	$opts->{passfail} = 1;
+	return bosh($opts,@args);
+}
+
+sub do_or_die {
+	my @args = @_;
+	my $opts = (ref($args[0]) eq 'HASH') ? shift @args : {};
+	delete @{$opts}{qw/interative passfail/};
+	$opts->{onfailure} = shift @args;
+	execute($opts,@args);
+	return undef;
+}
+
+sub get {
+	my @args = @_;
+	my $opts = (ref($args[0]) eq 'HASH') ? shift @args : {};
+	delete @{$opts}{qw/interative passfail/};
+	my $out = execute($opts,@args);
+	return $out
```

Comment:
If we modify execute() to return ($output, $rc) in list context, and $rc == 0 in scalar context, we can do this:

```
(my $out, undef) = execute($opts, @args);
return $out;
```